### PR TITLE
Implement a token endpoint for the REST API

### PR DESCRIPTION
**Pull request by SuKSW**
### Purpose

This PR will add the following:

- Implement token endpoint. Add REST validations. Refactor.
- Format config.toml and add param token private key
- Enable the commands below in apictl,
  - `apictl mg login dev -u admin -p admin`
  - `apictl mg deploy api -f petstore -e dev`
  - `apictl mg deploy api -f petstore -e dev -o` (overrides an existing)
  - `apictl mg get apis -e dev`
  - `apictl mg undeploy api -n SwaggerPetstore -v 1.0.5 -e dev`
  - `apictl mg logout dev`
  NOTE: the command `apictl mg add env prod --adapter https://localhost:9843` needs to be used before the above commands.

### Issues

Fixes https://github.com/wso2/product-microgateway/issues/1674, https://github.com/wso2/product-microgateway/issues/1444

### Automation tests
 - Unit tests added: Yes/No
 - Integration tests added: Yes/No

### Tested environments

Not Tested

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

**Comment by praminda** [comment]
Need a github issue for this. Please create one and link it under `Issues` heading. Also move the PR description under `Purpose` heading.

File: adapter/internal/api/restserver/config_restapi.go [hunk 1]
```diff
@@ -59,29 +61,61 @@ func configureAPI(api *operations.RestapiAPI) http.Handler {
...
-		principal *models.Principal) middleware.Responder {
+	api.BearerTokenAuth = func(token string, scopes []string) (*models.Principal, error) {
+		valid, err := auth.ValidateToken(token, scopes)
+		if err != nil || !valid {
+			logger.LoggerAPI.Info("The provided token is not valid or server error")
+			return nil, errors.Unauthenticated("The provided token is not valid or server error")
+		}
+		p := models.Principal{
+			Token:    token,
+			Tenant:   "xxxx",
```
Comment (praminda) [inline adapter/internal/api/restserver/config_restapi.go:88]: what does `xxxx` mean. can't this be nil or empty?

File: adapter/internal/api/restserver/config_restapi.go [hunk 2]
```diff
@@ -59,29 +61,61 @@ func configureAPI(api *operations.RestapiAPI) http.Handler {
...
 		}
 		logger.LoggerAPI.Debugf("Principal : %v", p)
 		return &p, nil
 	}
 
-	api.APIIndividualDeleteApisHandler = api_individual.DeleteApisHandlerFunc(func(params api_individual.DeleteApisParams,
-		principal *models.Principal) middleware.Responder {
+	api.BearerTokenAuth = func(token string, scopes []string) (*models.Principal, error) {
+		valid, err := auth.ValidateToken(token, scopes)
+		if err != nil || !valid {
```
Comment (praminda) [inline adapter/internal/api/restserver/config_restapi.go:82]: if `err` for an actual error case (adapter failed to process the token | server error) we need to make the response `500`. 401 should be only if the token was processed properly but invalid cases.

File: adapter/internal/mgw/mgw.go [hunk 3]
```diff
@@ -128,6 +129,12 @@ func runManagementServer(server xdsv3.Server, enforcerServer wso2_server.Server,
 			logger.LoggerMgw.Error(err)
 		}
 	}()
+
+	go func() {
+		if err = auth.Init(); err != nil {
+			logger.LoggerMgw.Error("error while initializing autherization component.", err)
```
Comment (praminda) [inline adapter/internal/mgw/mgw.go:135]:
```suggestion
			logger.LoggerMgw.Error("Error while initializing authorization component.", err)
```

File: resources/conf/config.toml [hunk 4]
```diff
@@ -1,52 +1,57 @@
...
-[[adapter.server.users]]
-username = "admin"
-password = "$env{adapter_admin_pwd}"
+[adapter] # --------------------------------------------------------
+
+# server: configuration related to REST API
+[adapter.server]
+  host = "0.0.0.0"
+  port = "9843"
+  tokenDuration = "1h"
```
Comment (praminda) [inline resources/conf/config.toml:9]: If we don't support `m` or `s` we shouldn't ask for that suffix. We should only expect a `int` value for the validity period in hours. We can add a comment saying `Token validity period in hours`.
Having `h` here is misleading and will cause users to try this with `m` and `s`

[the remark this answers is not in the document]
Comment (SuKSW) [reply resources/conf/config.toml:9]: Updating this to tokenTTL

File: adapter/internal/api/restserver/config_restapi.go (hunk 1 above)
Comment (SuKSW) [reply adapter/internal/api/restserver/config_restapi.go:88]: I am not exactly sure. It was there for basic auth, therefore I used the same with the token

File: adapter/internal/api/restserver/config_restapi.go (hunk 2 above)
Comment (SuKSW) [reply adapter/internal/api/restserver/config_restapi.go:82]: Correcting as mentioned.

File: adapter/internal/mgw/mgw.go (hunk 3 above)
Comment (SuKSW) [reply adapter/internal/mgw/mgw.go:135]: Updating this

File: resources/conf/config.toml (hunk 4 above)
Comment (SuKSW) [reply resources/conf/config.toml:9]: 1h is the default value. I not sure whether we decided not to support m.

Comment (dinusha92) [reply resources/conf/config.toml:9]: We might need to support m as well right? What we can do is, default unit it h, if not whatever the unit mentioned in the config.

**Comment by SuKSW** [comment]
Update readme

File: adapter/internal/api/restserver/config_restapi.go (hunk 1 above)
Comment (SuKSW) [reply adapter/internal/api/restserver/config_restapi.go:92]: remove tenant

File: adapter/internal/auth/auth.go [hunk 5]
```diff
@@ -26,3 +42,139 @@ func GetBasicAuth(username, password string) string {
 	auth := username + ":" + password
 	return base64.StdEncoding.EncodeToString([]byte(auth))
 }
+
+// ValidateCredentials checks whether the provided username and password are valid
+func ValidateCredentials(username, password string, config *config.Config) bool {
+	valid := false
+	for _, regUser := range config.Adapter.Server.Users {
+		if username == regUser.Username && password == regUser.Password {
+			valid = true
```
Comment (SuKSW) [inline adapter/internal/auth/auth.go:51]: return true

File: adapter/internal/auth/auth.go [hunk 6]
```diff
@@ -26,3 +42,139 @@ func GetBasicAuth(username, password string) string {
...
+		}
+	}
+	return valid
+}
+
+func validateUser(username string, config *config.Config) bool {
+	valid := false
+	for _, regUser := range config.Adapter.Server.Users {
+		if username == regUser.Username {
+			valid = true
```
Comment (SuKSW) [inline adapter/internal/auth/auth.go:61]: return true

File: adapter/internal/auth/auth.go [hunk 7]
```diff
@@ -26,3 +42,139 @@ func GetBasicAuth(username, password string) string {
...
+	privateKey, err := x509.ParsePKCS8PrivateKey(block.Bytes)
+	storedPrivateKey = privateKey.(*rsa.PrivateKey)
+	return err
+}
+
+func setTokenDuration(conf *config.Config) {
+	tokenDurationString := conf.Adapter.Server.TokenTTL
+	if tokenDurationString == "" {
+		loggers.LoggerAuth.Warn("Token duration not set. Set to default value: 1h")
+		tokenDurationString = "1h"
```
Comment (SuKSW) [inline adapter/internal/auth/auth.go:95]: comment time h m s functionality

File: adapter/internal/auth/auth.go [hunk 8]
```diff
@@ -26,3 +42,139 @@ func GetBasicAuth(username, password string) string {
...
+	token, err := jwt.ParseString(
+		accessToken,
+		jwt.WithValidate(true),
+		jwt.WithVerify(jwa.RS256, &privateKey.PublicKey),
+	)
+	if err != nil {
+		loggers.LoggerAPI.Errorf("Failed to parse JWT token: %s", err)
+		return false, nil
+	}
+	tokenUser, _ := token.Get("username")
```
Comment (SuKSW) [inline adapter/internal/auth/auth.go:152]: constants for username, scope

File: adapter/resources/adapterAPI.yaml [hunk 9]
```diff
@@ -115,24 +147,34 @@ paths:
           in: query
           type: string
           required: true
+          minLength: 3
+          maxLength: 255
+          pattern: ^[a-zA-Z0-9_~.-]*$
```
Comment (SuKSW) [inline adapter/resources/adapterAPI.yaml:152]: check open api v3. with a valid url for servers object.

File: resources/conf/config.toml (hunk 4 above)
Comment (SuKSW) [reply resources/conf/config.toml]: Adding a comment to the variable

File: README.md [hunk 10]
```diff
@@ -42,6 +42,10 @@ export PATH=$PATH:<CLI_HOME>
 ```
 
 3. Let's create our first project with name "petstore" by adding the [open API definition](https://petstore.swagger.io/v2/swagger.json) of the petstore . You can do that by executing the following command using your command line tool.
+
+  NOTE: If you have used a previous version of apictl before, remember to delete the directories
+ .wso2apictl and .wso2apictl.local that are located in `/home/<your-pc-username>`. Deleting them will make the newer apictl create them again, with content compatible with the current version.
+
```
Comment (dinusha92) [inline README.md:48]: @SuKSW let's mention here to keep a backup and remove the directory.

Comment (SuKSW) [reply README.md:48]: mentioning here to keep a backup and remove the directory.

File: adapter/internal/auth/auth.go (hunk 7 above)
Comment (SuKSW) [reply adapter/internal/auth/auth.go:100]: comment added to config file

**Review by dinusha92 — APPROVED** [review]

**Pull request merged**